### PR TITLE
[lld][ARM] Fix assertion when mixing ARM and Thumb objects

### LINKED ISSUE #99008: [LLD][ARM] Assertion "If the source is ARM, we should not need Thumb PLTs"
This reproducer tarball (found while building https://github.com/ARM-software/LLVM-embedded-toolchain-for-Arm/) causes an assertion in lld:

[repro.tar.gz](https://github.com/user-attachments/files/16246302/repro.tar.gz)

```
$ /work/et/build/llvm/bin/ld.lld --chroot . -l crt0-semihost -l semihost -rpath work/et/build/compiler-rt/armv7a_soft_nofp_exn_rtti/src/compiler_rt_armv7a_soft_nofp_exn_rtti-build/lib/arm-unknown-none-eabi tmp/ucmpdi2_test-71ac06.o work/et/build/compiler-rt/armv7a_soft_nofp_exn_rtti/src/compiler_rt_armv7a_soft_nofp_exn_rtti-build/lib/arm-unknown-none-eabi/libclang_rt.builtins.a -l c -l m -Bstatic -EL --script picolibcpp.ld -L work/et/build/compiler-rt/armv7a_soft_nofp_exn_rtti/src/compiler_rt_armv7a_soft_nofp_exn_rtti-build/lib/arm-unknown-none-eabi -L work/et/build/llvm/lib/clang-runtimes/arm-none-eabi/armv7a_soft_nofp_exn_rtti/lib -L work/et/build/compiler-rt/armv7a_soft_nofp_exn_rtti/src/compiler_rt_armv7a_soft_nofp_exn_rtti-build/lib/arm-unknown-none-eabi -L work/et/build/llvm/lib/clang-runtimes/arm-none-eabi/armv7a_soft_nofp_exn_rtti/lib --target2 rel -o ucmpdi2_test.c.tmp
ld.lld: /work/et/repos/llvm-project/lld/ELF/Arch/ARM.cpp:396: virtual bool {anonymous}::ARM::needsThunk(lld::elf::RelExpr, lld::elf::RelType, const lld::elf::InputFile*, uint64_t, const lld::elf::Symbol&, int64_t) const: Assertion `!config->armThumbPLTs && "If the source is ARM, we should not need Thumb PLTs"' failed.
PLEASE submit a bug report to https://github.com/ARM-software/LLVM-embedded-toolchain-for-Arm/issues and include the crash backtrace.
Stack dump:
0.      Program arguments: /work/et/build/llvm/bin/ld.lld --chroot . -l crt0-semihost -l semihost -rpath work/et/build/compiler-rt/armv7a_soft_nofp_exn_rtti/src/compiler_rt_armv7a_soft_nofp_exn_rtti-build/lib/arm-unknown-none-eabi tmp/ucmpdi2_test-71ac06.o work/et/build/compiler-rt/armv7a_soft_nofp_exn_rtti/src/compiler_rt_armv7a_soft_nofp_exn_rtti-build/lib/arm-unknown-none-eabi/libclang_rt.builtins.a -l c -l m -Bstatic -EL --script picolibcpp.ld -L work/et/build/compiler-rt/armv7a_soft_nofp_exn_rtti/src/compiler_rt_armv7a_soft_nofp_exn_rtti-build/lib/arm-unknown-none-eabi -L work/et/build/llvm/lib/clang-runtimes/arm-none-eabi/armv7a_soft_nofp_exn_rtti/lib -L work/et/build/compiler-rt/armv7a_soft_nofp_exn_rtti/src/compiler_rt_armv7a_soft_nofp_exn_rtti-build/lib/arm-unknown-none-eabi -L work/et/build/llvm/lib/clang-runtimes/arm-none-eabi/armv7a_soft_nofp_exn_rtti/lib --target2 rel -o ucmpdi2_test.c.tmp
 #0 0x0000561187063f00 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/work/et/build/llvm/bin/ld.lld+0x1c25f00)
 #1 0x000056118706131f llvm::sys::RunSignalHandlers() (/work/et/build/llvm/bin/ld.lld+0x1c2331f)
 #2 0x0000561187061475 SignalHandler(int) Signals.cpp:0:0
 #3 0x00007f1c1a242520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x00007f1c1a2969fc __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #5 0x00007f1c1a2969fc __pthread_kill_internal ./nptl/pthread_kill.c:78:10
 #6 0x00007f1c1a2969fc pthread_kill ./nptl/pthread_kill.c:89:10
 #7 0x00007f1c1a242476 gsignal ./signal/../sysdeps/posix/raise.c:27:6
 #8 0x00007f1c1a2287f3 abort ./stdlib/abort.c:81:7
 #9 0x00007f1c1a22871b _nl_load_domain ./intl/loadmsgcat.c:1177:9
#10 0x00007f1c1a239e96 (/lib/x86_64-linux-gnu/libc.so.6+0x39e96)
#11 0x00005611871dfce6 (anonymous namespace)::ARM::needsThunk(lld::elf::RelExpr, unsigned int, lld::elf::InputFile const*, unsigned long, lld::elf::Symbol const&, long) const ARM.cpp:0:0
#12 0x00005611872f6103 lld::elf::ThunkCreator::createThunks(unsigned int, llvm::ArrayRef<lld::elf::OutputSection*>) (/work/et/build/llvm/bin/ld.lld+0x1eb8103)
#13 0x00005611873df80d (anonymous namespace)::Writer<llvm::object::ELFType<(llvm::endianness)1, false>>::finalizeAddressDependentContent() Writer.cpp:0:0
#14 0x00005611873e0d33 (anonymous namespace)::Writer<llvm::object::ELFType<(llvm::endianness)1, false>>::finalizeSections() Writer.cpp:0:0
#15 0x00005611873e2a2c (anonymous namespace)::Writer<llvm::object::ELFType<(llvm::endianness)1, false>>::run() Writer.cpp:0:0
#16 0x00005611873e3abd void lld::elf::writeResult<llvm::object::ELFType<(llvm::endianness)1, false>>() (/work/et/build/llvm/bin/ld.lld+0x1fa5abd)
#17 0x00005611871c9a75 void lld::elf::LinkerDriver::link<llvm::object::ELFType<(llvm::endianness)1, false>>(llvm::opt::InputArgList&) (/work/et/build/llvm/bin/ld.lld+0x1d8ba75)
#18 0x00005611871d344b lld::elf::LinkerDriver::linkerMain(llvm::ArrayRef<char const*>) (/work/et/build/llvm/bin/ld.lld+0x1d9544b)
#19 0x00005611871d54c0 lld::elf::link(llvm::ArrayRef<char const*>, llvm::raw_ostream&, llvm::raw_ostream&, bool, bool) (/work/et/build/llvm/bin/ld.lld+0x1d974c0)
#20 0x0000561187065f70 lld::unsafeLldMain(llvm::ArrayRef<char const*>, llvm::raw_ostream&, llvm::raw_ostream&, llvm::ArrayRef<lld::DriverDef>, bool) (/work/et/build/llvm/bin/ld.lld+0x1c27f70)
#21 0x0000561186fa02a6 lld_main(int, char**, llvm::ToolContext const&) (/work/et/build/llvm/bin/ld.lld+0x1b622a6)
#22 0x0000561186fa064b main (/work/et/build/llvm/bin/ld.lld+0x1b6264b)
#23 0x00007f1c1a229d90 __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
#24 0x00007f1c1a229e40 call_init ./csu/../csu/libc-start.c:128:20
#25 0x00007f1c1a229e40 __libc_start_main ./csu/../csu/libc-start.c:379:5
#26 0x0000561186f9fae5 _start (/work/et/build/llvm/bin/ld.lld+0x1b61ae5)
fish: Job 1, '/work/et/build/llvm/bin/ld.lld…' terminated by signal SIGABRT (Abort)
```

### PR DESCRIPTION
**Pull request by ostannard**
Previously, we selected the Thumb2 PLT sequences if any input object is marked as not supporting the ARM ISA, which then causes assertion failures when calls from ARM code in other objects are seen. I think the intention here was to only use Thumb PLTs when the target does not have the ARM ISA available, signalled by no objects being marked as having it available. To do that we need to track which ISAs we have seen as we parse the build attributes, and defer the decision  about PLTs until all input objects have been parsed.

This bug was triggered by real code in picolibc, which have some versions of string.h functions built with Thumb2-only build attributes, so that they are compatible with v7-A, v7-R and v7-M.

Fixes #99008.

### CODE REVIEW

File: lld/ELF/Driver.cpp [hunk 1]
```diff
@@ -1886,6 +1887,14 @@ static void setConfigs(opt::InputArgList &args) {
   }
 }
 
+// Initialise config options which depend on information previously parsed from
+// input files, such as Arm build attributes.
+static void setConfigsPostParse() {
```
Comment (smithp35) [inline lld/ELF/Driver.cpp:1892]: Given that `config->armThumbPLTs` is only used within Arch/ARM.cpp and it is made up of only two booleans we could remove `config->armThumbPLTs` , this function and in Arch/ARM.cpp replace config->armThumbPLTs` with a function call to
```
static bool hasThumbPLT() {
  return config->armHasThumb2ISA && !config->armHasArmISA;
} 
```

The other bit of prior art we have in this area is the decision to write PAC/BTI PLT sequences. For that we had to delay until all the input objects had been written to. That ended up being done in a target specific `getAArch64TargetInfo()` but there is also `getARMTargetInfo()` that could be used.

**Review by smithp35 — COMMENTED** [review]
Change looks good to me. I've got a couple of suggestions that avoid the need to create `setConfigsPostParse`.

File: lld/ELF/Driver.cpp (hunk 1 above)
Comment (ostannard) [reply lld/ELF/Driver.cpp]: Done.

File: lld/test/ELF/arm-mixed-plts.s [hunk 2]
```diff
@@ -0,0 +1,44 @@
+# REQUIRES: arm
+
+# RUN: rm -rf %t && split-file %s %t
+# RUN: llvm-mc -filetype=obj -arm-add-build-attributes -triple=armv7a-none-linux-gnueabi %t/a.s -o %t1.o
+# RUN: llvm-mc -filetype=obj -arm-add-build-attributes -triple=armv7a-none-linux-gnueabi %t/b.s -o %t2.o
+# RUN: ld.lld -shared %t1.o %t2.o -o %t.so
+# RUN: llvm-objdump -d %t.so | FileCheck %s
+
+# Check that, when the input is a mixture of objects which can and cannot use
```
Comment (MaskRay) [inline lld/test/ELF/arm-mixed-plts.s:9]: In new tests, we use `## ` for non-CHECK-non-RUN comments.

File: lld/test/ELF/arm-mixed-plts.s [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+# the ARM ISA, we use the default ARM PLT sequences.
+
+# CHECK:      000101d0 <.plt>:
+# CHECK-NEXT: 101d0: e52de004      str     lr, [sp, #-0x4]!
```
Comment (MaskRay) [inline lld/test/ELF/arm-mixed-plts.s:13]: The leading addresses are not significant and are omitted for new tests

File: lld/test/ELF/arm-mixed-plts.s (hunk 2 above)
Comment (ostannard) [reply lld/test/ELF/arm-mixed-plts.s]: Done

File: lld/test/ELF/arm-mixed-plts.s (hunk 3 above)
Comment (ostannard) [reply lld/test/ELF/arm-mixed-plts.s]: Done

**Review by smithp35 — APPROVED** [review]
Thanks for the updates. Looking good on my side. Please leave some time for MaskRay to approve too.

**Review by MaskRay — APPROVED** [review]

**Pull request merged**